library/core/src/num/dec2flt/decimal.rs

```diff
@@ -0,0 +1,384 @@
+//! Arbitrary-precision decimal class for fallback algorithms.
+//!
+//! This is only used if the fast-path (native floats) and
+//! the Eisel-Lemire algorithm are unable to unambiguously
+//! determine the float.
+//!
+//! The technique used is "Simple Decimal Conversion", developed
+//! by Nigel Tao and Ken Thompson. A detailed description of the
+//! algorithm can be found in "ParseNumberF64 by Simple Decimal Conversion",
+//! available online: <https://nigeltao.github.io/blog/2020/parse-number-f64-simple.html>.
+
+use crate::fmt::{self, Debug};
+use crate::num::dec2flt::common::{is_8digits, parse_digits, ByteSlice, ByteSliceMut};
+
+#[derive(Clone)]
+pub struct Decimal {
+    /// The number of significant digits in the decimal.
+    pub num_digits: usize,
+    /// The offset of the decimal point in the significant digits.
+    pub decimal_point: i32,
+    /// If the sign of the float is negative.
+    pub negative: bool,
+    /// If the number of significant digits stored in the decimal is truncated.
+    pub truncated: bool,
+    /// Buffer of the raw digits, in the range [0, 9].
+    pub digits: [u8; Self::MAX_DIGITS],
+}
+
+impl Debug for Decimal {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Decimal")
+            .field("num_digits", &self.num_digits)
+            .field("decimal_point", &self.decimal_point)
+            .field("negative", &self.negative)
+            .field("truncated", &self.truncated)
+            .field("digits", &(&self.digits[..self.num_digits]))
+            .finish()
+    }
+}
+
+impl PartialEq for Decimal {
+    fn eq(&self, other: &Self) -> bool {
+        self.num_digits == other.num_digits
+            && self.decimal_point == other.decimal_point
+            && self.negative == other.negative
+            && self.truncated == other.truncated
+            && self.digits[..] == other.digits[..]
+    }
+}
+
+impl Eq for Decimal {}
+
+impl Default for Decimal {
+    fn default() -> Self {
+        Self {
+            num_digits: 0,
+            decimal_point: 0,
+            negative: false,
+            truncated: false,
+            digits: [0; Self::MAX_DIGITS],
+        }
```

Comment:
By derive this, do you mean creating my own, custom class for `digits`, say, `Digits` which wraps the array, and then implement `Default` in terms of that (since `Debug`, `PartialEq`, and `Eq` have been removed)?

Reply:
I mean
```
#[derive(Clone, Default)]
pub struct Decimal {
```
and then removing this whole impl block.

Reply:
Ah ok, I found out why this uses a custom default: `const_generics` can't yet be used with Default, for the reasons described [here](https://doc.rust-lang.org/src/core/array/mod.rs.html#369-391).

```rust
// The Default impls cannot be done with const generics because `[T; 0]` doesn't
// require Default to be implemented, and having different impl blocks for
// different numbers isn't supported yet.

macro_rules! array_impl_default {
    {$n:expr, $t:ident $($ts:ident)*} => {
        #[stable(since = "1.4.0", feature = "array_default")]
        impl<T> Default for [T; $n] where T: Default {
            fn default() -> [T; $n] {
                [$t::default(), $($ts::default()),*]
            }
        }
        array_impl_default!{($n - 1), $($ts)*}
    };
    {$n:expr,} => {
        #[stable(since = "1.4.0", feature = "array_default")]
        impl<T> Default for [T; $n] {
            fn default() -> [T; $n] { [] }
        }
    };
}

array_impl_default! {32, T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T T}
```

Reply:
Yeah, impls can't be feature-gated - the manual impl is fine.